### DISCUSSION TITLE
Cancel a request if it takes longer than T seconds

### DISCUSSION THREAD

**Discussion by raqibhayder** [fastapi/fastapi#3501] [Questions] [answered]
In the application below, I want to cancel the request and throw a 5xx if it takes more than T seconds to execute. Is there an elegant way of doing this from the application layer using FastAPI or Starlette? 

``` python 
from fastapi import FastAPI
from time import sleep

app = FastAPI()

@app.get("/slow")
async def slow():
    # Do a long task here which takes > T seconds

🙏

**Comment by HarrySky** [comment, marked as answer]
You can use `asyncio.wait_for` (https://docs.python.org/3/library/asyncio-task.html#asyncio.wait_for) and in case of `asyncio.TimeoutError` just return 500 response

**Comment by ghandic** [comment]
As @HarrySky suggests you can use `asyncio.wait_for()`, but only if your long-running task is async

```python
import time
import asyncio

from fastapi import FastAPI

app = FastAPI()


async def eternityish():
    # Sleep for one hour
    await asyncio.sleep(3600)
    print("yay!")


async def eternity():
    # Sleep for one hour
    time.sleep(3600)
    print("yay!")


@app.get("/slow")
async def slow():
    # Do a long task here which takes > T seconds
    T = 5.0
    await asyncio.wait_for(eternityish(), timeout=T)


@app.get("/slow2")
async def slow():
    # Do a long task here which takes > T seconds
    T = 5.0
    await asyncio.wait_for(eternity(), timeout=T)
```

Does @HarrySky 's solution answer your question? If so could you mark it as answered 😄
